### ISSUE TITLE
Shiny Download Button

### ISSUE THREAD

**Issue by Chuck-Field** [att/rcloud.shiny#4]
The "Download Button" for shiny does not initiate a file download and instead opens another instance of the same Shiny app, in a second tab.  I provided sample code, directly from the Shiny Gallery that was modified for RCloud as an example. 
It would be nice to be able to download tables, through the browser, but this does not seem to work in RCloud. And to be honest, I don't know if this is bugged code on my part, missing code I need to add or something wrong on the RCloud end. I do know I've tried to use the button on different projects and I get the same results in RCloud. Thanks for looking into this. 

``` r
library(rcloud.shiny)
rcloud.shinyApp(
ui = bootstrapPage(
  titlePanel('Downloading Data'),
  sidebarLayout(
    sidebarPanel(
      selectInput("dataset", "Choose a dataset:", 
                  choices = c("rock", "pressure", "cars")),
      downloadButton('downloadData', 'Download')
    ),
    mainPanel(
      tableOutput('table')
 ))),

shinyServer(function(input, output) {
      datasetInput <- reactive({
    switch(input$dataset,
           "rock" = rock,
           "pressure" = pressure,
           "cars" = cars)
  })
  output$table <- renderTable({
    datasetInput()
  })
  output$downloadData <- downloadHandler(
    filename = function() { 
         paste(input$dataset, '.csv', sep='') 
     },
    content = function(file) {
      write.csv(datasetInput(), file)
    }
  )
}))
```

**Comment by gordonwoodhull** [comment]
This is another known compatibility issue, like #5. IIRC this requires an Ajax call from the client to the server, which we do not yet support. We only currently support the websocket communication between the Shiny client and server.

Thanks for reporting this!

**Comment by gordonwoodhull** [comment]
Er, probably a simple http request, not Ajax. I think the timer issue requires Ajax. (What timer issue, you ask? I will report it when I find an old email about it.)

**Comment by gordonwoodhull** [comment]
Here is a notebook with a temporary workaround: https://rcloud.social/edit.html?notebook=2b2bcfe5bfb5fa9e98d2274604a1f3df

It's a function you can call from your shiny code to invoke a download.

Hopefully we will have a new, more compatible version of rcloud.shiny in the coming weeks.

**Comment by gordonwoodhull** [comment]
RCloud 1.6.1 will have a built-in API `rcloud.download.file` with the same code - att/rcloud#2005.

**Comment by gordonwoodhull** [comment]
I'm fairly confident that we'll have full Shiny compatibility through this rcloud.shiny library (with no special cases for anything), via iframes, before the end of the year.

**Comment by Chuck-Field** [comment]
That's great news. The sooner the better. Having that functionality will enable me to create shiny apps, that my non-RCloud customers can take advantage of.  Pushing and pulling data is the biggest road block I have when getting non-techie people engaged. They need to be able to upload, process, and download all in Shiny through point and click decisions.

**Comment by gordonwoodhull** [comment]
@Chuck-Field, I take it you are unable to use the workaround? I agree that it's better to go the pure Shiny route than to have some Frankenapp mixing RCloud and Shiny methods, but I feel that I've demonstrated a way forward and don't think you should be blocked at this point.

**Comment by Chuck-Field** [comment]
We have some work around processes in place. We have a SQL server that we can easily push and pull data in and out of RCloud and we built some Excel (VBA) processes for our people to push and pull data from the SQL Server.  That solution works fine for what we are doing today, but I believe we can get closer to "real time" solutions on ad hoc needs, giving our customers the ability to push and pull data from their desktops.  This example, we are using today, is the culmination of presentation app which pulls data from the SQL server,
https://rcloud.web.att.com/shiny.html?notebook=3c5cf6995a79b7864e02
https://rcloud.web.att.com/edit.html?notebook=3c5cf6995a79b7864e02
, a shiny app that pushes "text data" to our server, and we have an excel template that users push "numeric data" to the server closing the loop on the presentation app in shiny that is pretty much hands off for us developing.

**Comment by gordonwoodhull** [comment]
Fixed in rcloud.shiny 0.4 (RCloud 1.7)
